src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -300,112 +302,122 @@ private static void BindProperty(PropertyInfo property, object instance, IConfig
+                    Debug.Assert(bindingPoint.Value is not null);
 
-                    if (interfaceGenericType is not null &&
-                        (interfaceGenericType == typeof(ICollection<>) || interfaceGenericType == typeof(IList<>)))
+                    // At this point we know that we have a non-null bindingPoint.Value, we just have to populate the items
+                    // using the IDictionary<> or ICollection<> interfaces, or properties using reflection.
+                    Type? dictionaryInterface = FindOpenGenericInterface(typeof(IDictionary<,>), type);
+
+                    if (dictionaryInterface != null)
                     {
-                        // For ICollection<T> and IList<T> we bind them to mutable List<T> type.
-                        Type genericType = typeof(List<>).MakeGenericType(type.GenericTypeArguments[0]);
-                        bindingPoint.SetValue(Activator.CreateInstance(genericType));
+                        BindDictionary(bindingPoint.Value, dictionaryInterface, config, options);
                     }
                     else
                     {
-                        bindingPoint.SetValue(CreateInstance(type, config, options));
+                        Type? collectionInterface = FindOpenGenericInterface(typeof(ICollection<>), type);
+                        if (collectionInterface != null)
+                        {
+                            BindCollection(bindingPoint.Value, collectionInterface, config, options);
+                        }
+                        else
+                        {
+                            BindProperties(bindingPoint.Value, config, options);
+                        }
                     }
                 }
-
-                Debug.Assert(bindingPoint.Value is not null);
-
-                // At this point we know that we have a non-null bindingPoint.Value, we just have to populate the items
-                // using the IDictionary<> or ICollection<> interfaces, or properties using reflection.
-                Type? dictionaryInterface = FindOpenGenericInterface(typeof(IDictionary<,>), type);
-
-                if (dictionaryInterface != null)
-                {
-                    BindDictionary(bindingPoint.Value, dictionaryInterface, config, options);
-                }
                 else
                 {
-                    Type? collectionInterface = FindOpenGenericInterface(typeof(ICollection<>), type);
-                    if (collectionInterface != null)
+                    if (isParentCollection)
                     {
-                        BindCollection(bindingPoint.Value, collectionInterface, config, options);
-                    }
-                    else
-                    {
-                        BindProperties(bindingPoint.Value, config, options);
+                        bindingPoint.SetValue(CreateInstance(type, config, options));
```

Comment:
>SetValue [](http://example.com/codeflow?start=37&length=8)

This should be `TrySetValue` to avoid the case when having the property is readonly..

Reply:
Changed by amend: https://github.com/dotnet/runtime/pull/86485/commits/a8944dd781fa05a90f4850ae990b6904d75d38ca#diff-2e6f75e1d4f577e1c8c1b6a32d158ce7812a38ad8fcd3b4d762157b92c03e890R420